Removes M365 group and connected site. Closes #5224

Saurabh7019 opened the pull request:
Closes #5224

milanholemans commented:
Thank you @Saurabh7019, we'll try to review it ASAP!

Saurabh7019 commented:
Oh! I did git rebase main to resolve the conflicts and now all my changes are gone. :( Is there a way to recover the lost changes? @milanholemans

milanholemans commented:
Hi @Saurabh7019, is this the commit you are looking for? https://github.com/pnp/cli-microsoft365/commit/5fdba369cb261ed46e0e9889959125606495ea2c

If so, I suggest that you run:

- `git reset main --hard` (start again from the main branch)
- `git cherry-pick 5fdba369cb261ed46e0e9889959125606495ea2c` and resolve merge conflicts if there are any (apply the specific commit back to the branch)
- `git push --force` (rewrite history)


Hopefully that will do it 🤞

Saurabh7019 commented:
Thank you for saving me a lot of time; I will update this PR tonight!

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.ts:
```diff
@@ -115,6 +109,124 @@ class AadM365GroupRemoveCommand extends GraphCommand {
+
+  private async deleteM365GroupFromRecyclebin(logger: Logger, id: string): Promise<void> {
+    const maxRetries = 10;
+    const intervalInMs: number = 6000;
```
Let's either use number in both consts or none to keep it consistent

waldekmastykarz replied:
Let's mention these values in the command's docs to properly manage users' expectations

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.ts:
```diff
@@ -115,6 +109,124 @@ class AadM365GroupRemoveCommand extends GraphCommand {
+      }
+    }
+
+    await logger.logToStderr(`Group could not be removed from the recycle bin after all retries.`);
```
This message is confusing. If the deleting process is still in pending due to extra server load, it could give the impression that we haven't even started deleting, while in fact it's in progress. Let's ensure that we make it clear that the process is pending but haven't finished before the assumed number of retries

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.ts:
```diff
@@ -115,6 +109,124 @@ class AadM365GroupRemoveCommand extends GraphCommand {
+      }
+      else {
+        if (this.verbose) {
+          await logger.logToStderr(`Group has not been deleted yet. Waiting and retrying...`);
```
Let's communicate the duration of waiting and the number of attempts left

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.ts:
```diff
@@ -115,6 +109,124 @@ class AadM365GroupRemoveCommand extends GraphCommand {
+
+    try {
+      const response = await request.get<{ id: string }>(requestOptions);
+      return Boolean(response && response.id);
```
Do we need to call `Boolean()`? Won't `response && response.id` already return a bool value?

Saurabh7019 replied:
Not having explicit boolean conversion is throwing type mismatch error. response && response.id may not always provide the expected boolean result, especially if response is undefined.

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.ts:
```diff
@@ -115,6 +109,124 @@ class AadM365GroupRemoveCommand extends GraphCommand {
+    }
+  }
+
+  private async sleep(logger: Logger, ms: number): Promise<void> {
```
This seems like a util function that we should have available centrally to avoid repetition. Let's add it to utils

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.ts:
```diff
@@ -115,6 +109,124 @@ class AadM365GroupRemoveCommand extends GraphCommand {
+
+    const res: FormDigestInfo = await spo.ensureFormDigest(this.spoAdminUrl as string, logger, undefined, this.debug);
+
+    const requestOptions: any = {
```
Previously, we defined request options as `CliRequestOptions`. Let's stay consistent and use it here too

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.ts:
```diff
@@ -115,6 +109,124 @@ class AadM365GroupRemoveCommand extends GraphCommand {
+      data: `<Request AddExpandoFieldTypeSuffix="true" SchemaVersion="15.0.0.0" LibraryVersion="16.0.0.0" ApplicationName="${config.applicationName}" xmlns="http://schemas.microsoft.com/sharepoint/clientquery/2009"><Actions><ObjectPath Id="185" ObjectPathId="184" /><Query Id="186" ObjectPathId="184"><Query SelectAllProperties="false"><Properties><Property Name="IsComplete" ScalarProperty="true" / ...
+    };
+
+    await request.post(requestOptions);
```
Don't we need to handle an error response here?

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.ts:
```diff
@@ -115,6 +109,124 @@ class AadM365GroupRemoveCommand extends GraphCommand {
+    const requestOptions: CliRequestOptions = {
+      url: `${this.resource}/v1.0/directory/deletedItems/${id}`,
+      headers: {
+        'accept': 'application/json;odata.metadata=none'
```
We don't need to surround `accept` in quotes

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.ts:
```diff
@@ -115,6 +109,124 @@ class AadM365GroupRemoveCommand extends GraphCommand {
+    const requestOptions: CliRequestOptions = {
+      url: `${this.resource}/v1.0/directory/deletedItems/${id}`,
+      headers: {
+        'accept': 'application/json;odata.metadata=none'
```
We don't need to surround `accept` with quotes

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.ts:
```diff
@@ -115,6 +109,124 @@ class AadM365GroupRemoveCommand extends GraphCommand {
+    const intervalInMs: number = 6000;
+
+    for (let retries = 0; retries < maxRetries; retries++) {
+      if (await this.isM365GroupInDeletedItemsList(id)) {
```
Do we need to check this on each call and resubmit the delete operation each time? Resubmitting deletion multiple times seems unnecessary. Shouldn't we submit it only once and then wait until the site disappears from deleting items (assuming that's the only way to check if it's actually been deleted)?

Saurabh7019 replied:
The code does not resubmit the delete operation each time. It checks if the M365 group is in the deleted items list on each iteration, and if the group is found, it proceeds to remove it permanently and then exits the loop. There is no resubmission of the delete operation beyond this point.

I fear I may not have fully understood the comment. Could you please let me know how I can optimize it further?

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.ts:
```diff
@@ -115,6 +109,124 @@ class AadM365GroupRemoveCommand extends GraphCommand {
+    await request.post(requestOptions);
+  }
+
+  private async deleteM365GroupFromRecyclebin(logger: Logger, id: string): Promise<void> {
```
Let's update the spelling of _recycle bin_ as `RecycleBin` as it's two words, and align it with the `deleteSiteFromRecycleBin` method where we spell it correctly

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.ts:
```diff
@@ -115,6 +109,124 @@ class AadM365GroupRemoveCommand extends GraphCommand {
+    const requestOptions: CliRequestOptions = {
+      url: `${this.spoAdminUrl}/_api/GroupSiteManager/Delete?siteUrl='${url}'`,
+      headers: {
+        'content-type': 'application/json;odata=nometadata',
```
Do we need to set `content-type` even though we're not submitting anything?

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.ts:
```diff
@@ -115,6 +109,124 @@ class AadM365GroupRemoveCommand extends GraphCommand {
+    }
+
+    const spoAdminUrl = await spo.getSpoAdminUrl(logger, this.debug);
+    this.spoAdminUrl = spoAdminUrl;
```
Since we need this URL in multiple methods, I'd rather not retrieve it inside this method, but take it a level up. Otherwise, if ever changed the execution order, it could lead to issues, so let's get ahead of it. Also, let's avoid the unnecessary assignment to a local const and then to a class property, and assign to class property directly

Saurabh7019 replied:
Moved the function call a level up, removed the class property, and passed it in functions to keep it consistent with siteUrl.

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.ts:
```diff
@@ -115,6 +109,124 @@ class AadM365GroupRemoveCommand extends GraphCommand {
     }
   }
+
+  private async getM365GroupSiteURL(logger: Logger, id: string): Promise<string> {
```
As we're spelling URL as `Url` in other variables, let's align this name consistently as `getM365GroupSiteUrl`

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.ts:
```diff
@@ -115,6 +109,124 @@ class AadM365GroupRemoveCommand extends GraphCommand {
+    };
+
+    const res = await request.get<{ webUrl: string }>(requestOptions);
+    return res.webUrl.substring(0, res.webUrl.lastIndexOf('/'));
```
What's the point of the substring: do we want to trim the trailing `/` or remove the web name from the URL? If it's the former, let's use `.trim()` instead of `substring()`. If it's the latter, let's clarify that with a comment.

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.ts:
```diff
@@ -69,29 +74,18 @@ class AadM365GroupRemoveCommand extends GraphCommand {
-        await logger.logToStderr(`Removing Microsoft 365 Group: ${args.options.id}...`);
-      }
+    if (this.verbose) {
+      await logger.logToStderr(`Removing Microsoft 365 Group: ${args.options.id}...`);
```
We placed this comment inside the `removeGroup` function so that it's displayed only after the user confirmed that they want to delete the group. Let's keep it that way and move it back to the function.

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.ts:
```diff
@@ -69,29 +74,18 @@ class AadM365GroupRemoveCommand extends GraphCommand {
-
-        await request.delete(requestOptions);
+        const siteUrl = await this.getM365GroupSiteURL(logger, args.options.id);
+        await this.deleteM365GroupSite(logger, siteUrl);
```
Am I understanding it correctly, that deleting the group site will also delete the group? If so, let's clarify this in a comment

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.spec.ts:
```diff
@@ -11,21 +11,70 @@ import { pid } from '../../../../utils/pid.js';
   let commandInfo: CommandInfo;
   let promptOptions: any;
 
+  const groupId = "3e6e705d-6fb5-4ca7-84dc-3c8f5154fe2c";
```
Let's use single quotes for consistency

waldekmastykarz commented on src/m365/aad/commands/m365group/m365group-remove.spec.ts:
```diff
@@ -11,21 +11,70 @@ import { pid } from '../../../../utils/pid.js';
     sinon.stub(telemetry, 'trackEvent').returns();
     sinon.stub(pid, 'getProcessName').returns('');
     sinon.stub(session, 'getId').returns('');
+    sinon.stub(global, 'setTimeout').callsFake((fn) => {
```
As we're stubbing `setTimeout` here, we should restore it in `after`

waldekmastykarz reviewed (changes requested):
Nice start. Let's do some adjustments before we continue

Saurabh7019 commented:
I am unable to drop the 'Empty commit' that I added to re-trigger the build pipeline. For now, I will mark the PR as 'Ready for review' and attempt to remove the empty commit later.

Adam-it commented:
@waldekmastykarz I saw you already started this one. Do you want to continue?

milanholemans commented on src/utils/timeUtil.ts:
```diff
@@ -0,0 +1,5 @@
+export const timeUtil = {
+  sleep(ms: number): Promise<void> {
+    return new Promise(resolve => setTimeout(resolve, ms));
+  }
+};
```
A tip, instead of doing this, you could use an out of the box async timeout function.

```ts
import { setTimeout } from 'timers/promises';
await setTimeout(500);
```

milanholemans reviewed (changes requested):
Haven't done a full review, just noticed something we can improve.

Saurabh7019 commented:
> Haven't done a full review, just noticed something we can improve.

I have removed the timeUtil and implemented the suggested change. Additionally, I have updated my previous email on the contributors list to reflect my current status as part of this PR. I hope that is okay.

Adam-it commented:
@Saurabh7019 sorry for the hold up. Do you think you could rebase before we proceed?

Adam-it reviewed (approved):
tested locally ✅
works nicely 👍
LGTM 👍
you rock 👏

Adam-it commented:
ready to merge 🚀

Adam-it commented:
merged manually. Awesome work 👏
